Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm

```diff
@@ -1333,4 +1333,92 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+{
+    auto blockPtr = makeBlockPtr([weakThis = WeakPtr { *this }] {
+        NSString *userInstalledFontsPath = [NSHomeDirectory() stringByAppendingPathComponent:@"Library/Fonts"];
+
+        NSDirectoryEnumerator<NSString *> *enumerator = [NSFileManager.defaultManager enumeratorAtPath:userInstalledFontsPath];
+
+        Vector<URL> fontURLs;
+        for (NSString *font in enumerator) {
+            if ([font hasSuffix:@"ttf"] || [font hasSuffix:@"ttc"] || [font hasSuffix:@"otf"]) {
+                NSString *fontsPath = [userInstalledFontsPath stringByAppendingPathComponent:font];
+                auto fontURL = URL::fileURLWithFileSystemPath(String(fontsPath));
+                fontURLs.append(fontURL);
+                RELEASE_LOG(Process, "Registering font url %s", fontURL.string().utf8().data());
+            }
+        }
+
+        RunLoop::protectedMain()->dispatch([weakThis, fontURLs = WTFMove(fontURLs)] {
+            if (RefPtr protectedThis = weakThis.get()) {
+                for (auto& process : protectedThis->m_processes)
+                    process->send(Messages::WebProcess::RegisterAdditionalFonts(additionalFonts(fontURLs)), 0);
+                protectedThis->m_additionalFontURLs = WTFMove(fontURLs);
+            }
+        });
+    });
+
+    if (m_additionalFontURLs) {
+        process.send(Messages::WebProcess::RegisterAdditionalFonts(additionalFonts(*m_additionalFontURLs)), 0);
+        return;
+    }
+    static bool hasStartedCachingUserInstalledFonts = false;
+    if (!hasStartedCachingUserInstalledFonts) {
+        dispatch_async(dispatch_get_global_queue(QOS_CLASS_DEFAULT, 0), blockPtr.get());
+        hasStartedCachingUserInstalledFonts = true;
+    }
+}
+
+void WebProcessPool::registerAssetFont(ASCIILiteral fontName, WebProcessProxy& process)
+{
+    auto nsFontName = fontName.createNSString();
+    NSDictionary *attributes = @{ (NSString *)kCTFontFamilyNameAttribute: nsFontName.get(), (NSString *)kCTFontRegistrationScopeAttribute: @(kCTFontPriorityComputer) };
+    auto description = adoptCF(CTFontDescriptorCreateWithAttributes((__bridge CFDictionaryRef)attributes));
+    NSArray *descriptions = [NSArray arrayWithObject:(__bridge id)description.get()];
+
+    auto blockPtr = makeBlockPtr([nsFontName, weakThis = WeakPtr { *this }](CTFontDescriptorMatchingState state, CFDictionaryRef progressParameter) {
+        if (state == kCTFontDescriptorMatchingDidFinish) {
+            RELEASE_LOG(Process, "Font matching finished, progress parameter = %@", (__bridge id)progressParameter);
+            RunLoop::protectedMain()->dispatch([weakThis, nsFontName] {
+                if (RefPtr protectedThis = weakThis.get()) {
+                    auto font = adoptCF(CTFontCreateWithName((__bridge CFStringRef)nsFontName.get(), 0.0, nullptr));
+                    auto cfFontURL = adoptCF(static_cast<CFURLRef>(CTFontCopyAttribute(font.get(), kCTFontURLAttribute)));
```

Comment:
`RetainPtr`

Reply:
Actually, if it is only used to construct a `WTF::URL`, we should just construct the `WTF::URL` right away and avoid the local cf variable.